internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java

```diff
@@ -0,0 +1,748 @@
+      return StatsErrorCode.LAG_MEASUREMENT_FAILURE.code;
+    }
+
+    updateCacheAsync(
+        pubSubTopicPartition,
+        cachedValue,
+        lastProducerTimestampCache,
+        () -> getProducerTimestampOfLastDataMessageWithRetriesAsync(
+            pubSubTopicPartition,
+            DEFAULT_MAX_RETRIES_FOR_POPULATING_TMD_CACHE_ENTRY));
+
+    return cachedValue.getValue();
+  }
+
+  /**
+   * This method retrieves last {@code lastRecordsCount} records from a topic partition and there are 4 steps below.
+   *  1. Find the current end offset N
+   *  2. Seek back {@code lastRecordsCount} records from the end offset N
+   *  3. Keep consuming records until the last consumed offset is greater than or equal to N
+   *  4. Return all consumed records
+   *
+   * There are 2 things to note:
+   *   1. When this method returns, these returned records are not necessarily the "last" records because after step 2,
+   *      there could be more records produced to this topic partition and this method only consume records until the end
+   *      offset retrieved at the above step 2.
+   *
+   *   2. This method might return more than {@code lastRecordsCount} records since the consumer poll method gets a batch
+   *      of consumer records each time and the batch size is arbitrary.
+   */
+  List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> consumeLatestRecords(
+      PubSubTopicPartition pubSubTopicPartition,
+      int lastRecordsCount) {
+    if (lastRecordsCount < 1) {
+      throw new IllegalArgumentException(
+          "Last record count must be greater than or equal to 1. Got: " + lastRecordsCount);
+    }
+    validateTopicPartition(pubSubTopicPartition);
+    PubSubConsumerAdapter pubSubConsumerAdapter = acquireConsumer();
+    boolean subscribed = false;
+    try {
+      // find the end offset
+      Map<PubSubTopicPartition, Long> offsetMap = pubSubConsumerAdapter.endOffsets(
+          Collections.singletonList(pubSubTopicPartition),
+          PUBSUB_OFFSET_API_TIMEOUT_DURATION_DEFAULT_VALUE);
+      if (offsetMap == null || offsetMap.isEmpty()) {
+        throw new VeniceException("Failed to get the end offset for topic-partition: " + pubSubTopicPartition);
+      }
+      long latestOffset = offsetMap.get(pubSubTopicPartition);
+      if (latestOffset <= 0) {
+        return Collections.emptyList(); // no records in this topic partition
+      }
+
+      // find the beginning offset
+      long earliestOffset =
+          pubSubConsumerAdapter.beginningOffset(pubSubTopicPartition, PUBSUB_OFFSET_API_TIMEOUT_DURATION_DEFAULT_VALUE);
+      if (earliestOffset == latestOffset) {
+        return Collections.emptyList(); // no records in this topic partition
+      }
+
+      // consume latest records
+      long consumePastOffset = Math.max(Math.max(latestOffset - lastRecordsCount, earliestOffset) - 1, -1);
+      pubSubConsumerAdapter.subscribe(pubSubTopicPartition, consumePastOffset);
```

Comment:
Since this class is using shared consumer here, shall we check whether the returned consumer has subscribed any other topic partitions or not?
Maybe it is a good practice to reset the consumer after acquiring the consumer from the consumer pool.

Reply:
Currently, any `subscribe` action performed on these shared consumers will always be followed by an `unsubscribe`. When we obtain a consumer, it should not have any existing subscriptions. If it does, it indicates a problem with this code/system.

We can adopt super defensive approach as you suggested, however.

Reply:
yeah, `unsubscribe` can be skipped if some unexpected error happen.

Reply:
Update with defensive check